src/runtime/qurt_hvx.cpp

```diff
@@ -52,4 +53,214 @@ WEAK void halide_qurt_hvx_unlock_as_destructor(void *user_context, void * /*obj*
+
+#if 0   // TODO: Opt: Box collapse disabled for now - needs more testing, and
+        // TODO       collapse candidates tend to exceed the stride mask size
+    // For boxes with dimension > 2 try to "collapse" any unit height
+    // dimensions by increasing the descriptor stride
+    uint32_t newdim = boxdim;
+    while (newdim > 2) {
+        uint32_t newpstride = 0;
+        if (pheight == 1) {     // if height is currently 1
+            newpstride = stride[iterdim-1] * elem_size;  // update stride
+        } else {
+            break;              // otherwise, we're done collapsing
+        }
+        if (newpstride == (newpstride & MASK16)) {  // and if it fits in mask...
+            pstride = newpstride;             // ...accept new stride
+            pheight = extent[iterdim];        // ...and height
+        } else {
+            break;              // otherwise, we're done collapsing
+        }
+        newdim--;       // remaining non-collapsed dimensions
+        iterdim++;      // update innermost iterate dimension
+    }
+    DBG_PREFETCH("  iterdim:%d\n", iterdim);
+#endif
+
+    pdir    = pdir & 0x1;           // bit  48
+    pstride = MIN(pstride, MASK16); // bits 47:32
+    pwidth  = MIN(pwidth,  MASK16); // bits 31:16
+    pheight = MIN(pheight, MASK16); // bits 15:0
+
+    uint64_t pdir64    = pdir;
+    uint64_t pstride64 = pstride;
+    uint64_t pdesc = (pdir64<<48) | (pstride64<<32) | (pwidth<<16) | pheight;
+
+    const uint32_t pbytes = pwidth * pheight;   // Bytes in prefetch descriptor
+    uint32_t tbytes = 0;                        // Total bytes prefetched
+    uint32_t tcnt = 0;                          // Total count of prefetch ops
+
+    // Hard coded descriptors for testing
+    // uint64_t pdesc = 0x1020002000002; // col, 512 width, 2 rows
+    // uint64_t pdesc = 0x1020002000001; // col, 512 width, 1 row
+    // uint64_t pdesc = 0x1144014400001; // col, 5184 width, 1 row
+
+    DBG_PREFETCH("  prefetch pdir:0x%x pstride:0x%x pwidth:0x%x pheight:0x%x\n",
+                           pdir, pstride, pwidth, pheight);
+    DBG_PREFETCH("  prefetch addr:0x%p pdesc:0x%llx\n", addr, pdesc);
+
+    // If iterdim extent is unity then move to next higher dimension
+    while ((iterdim < dim) && (extent[iterdim] == 1)) {
+        iterdim++;
+    }
+    DBG_PREFETCH("  iterdim:%d\n", iterdim);
+
+    if (iterdim >= dim) {       // No dimension remaining to iterate over
+
+        // Perform a single prefetch
+        DBG_PREFETCH("  l2fetch(0x%p, 0x%llx): %d bytes\n", addr, pdesc, pbytes);
+        __asm__ __volatile__ ("l2fetch(%0,%1)" : : "r"(addr), "r"(pdesc));
+        tbytes += pbytes;
+        tcnt++;
+
+    } else {    // Iterate for higher dimension boxes
+
+        // Get iteration stride and extents
+        int32_t iterstride = stride[iterdim] * elem_size;
+        int32_t iterextent = extent[iterdim];
+        iterextent = MIN(iterextent, MAX_PREFETCH);   // Limit # of prefetches
```

Comment:
I see now. I think it would be helpful to make a function that wraps the __asm__ blocks that constructs the pdesc from a human readable set of parameters. I think that would make this code easier to understand.